CEXT-6469 Admin UI SDK V2 (#579)
AdobeDocs/commerce-extensibility: src/pages/app-management/installation/admin-ui-sdk.md

---
title: Admin UI SDK configuration
description: Configure the Admin UI SDK menu, grid columns, mass actions, and order view buttons in the app.commerce.config file for App Management
keywords:
  - App Builder
  - Extensibility
  - App Management
  - Admin UI SDK
---

# Admin UI SDK configuration

The `adminUi` field in `app.commerce.config` defines how your app extends the Adobe Commerce Admin UI through the `commerce/backend-ui/2` extension point. It is not compatible with the `commerce/backend-ui/1` extension point used by the previous version of the Admin UI SDK.

At a high level, it is the top-level config block for declaring:

* Menu entries
* Grid columns
* Mass actions
* Order view buttons

> `commerce/backend-ui/2` reads the registration directly from the generated `app-config` runtime action, so no dedicated registration action is generated. When `adminUi` is defined, `init` and `generate all` automatically wire up the extension point, including the `pre-app-build` hook and the `workerProcess` declarations in `ext.config.yaml`. View-based features (a menu, a `view` mass action, or a `view` order view button) also get a minimal `web-src/` scaffold the first time they're added, using `.tsx` files for TypeScript configs and `.jsx` files otherwise.

Reviewer: Idem here in `no dedicated registration action` is generated.

## Add mass actions

Add mass actions that run against a selection of grid rows. Each entry uses `type` to select how it runs. When the value is `view`, it opens an iframe at `path`; when the value is `worker`, it invokes a `runtimeAction`:

```js
adminUi: {
  order: {
    massActions: [
      {
        type: "view",
        id: "export-orders",
        label: "Export",
        title: "Export orders",
        path: "#/export-orders",
        sandboxPermissions: ["allow-downloads"],
      },
      {
        type: "worker",
        id: "bulk-approve",
        label: "Approve",
        confirm: { message: "Approve the selected orders?" },
        runtimeAction: "orders/bulk-approve",
        timeout: 15,
        notifications: {
          success: "Orders approved.",
          error: "Approval failed. Check the runtime logs.",
        },
      },
    ],
  },
}
```

> Mass actions are supported on `order`, `product`, and `customer`. The `id` is authored as a bare name (for example `bulk-approve`); Commerce handles prefixing and collision resolution when rendering the final Admin UI configuration.

Reviewer 2: ```suggestion
Mass actions are supported on `order`, `product`, and `customer`. The `id` is authored as a bare name (for example `bulk-approve`). Commerce handles prefixing and collision resolution when rendering the final Admin UI configuration.
```

        path: "#/delete-order",
        level: 0,
        sortOrder: 80,
        sandboxPermissions: ["allow-modals", "allow-popups"],
        confirm: { message: "Are you sure you want to delete this order?" },
      },
      {
        type: "worker",
        id: "sync-inventory",
        label: "Sync inventory",
        description: "Pushes the latest stock counts for this order's items to the ERP.",
        runtimeAction: "orders/sync-inventory",
        timeout: 15,
        level: 1,
        sortOrder: 10,
        notifications: {
          success: "Inventory synced successfully.",
          error: "Inventory sync failed. Check the runtime logs.",
        },
      },
    ],
  },
}
```

Order view buttons are only available on `order`.

### Field applicability by variant

| Field | Common | `view` only | `worker` only |
|-------|:------:|:-----------:|:-------------:|
| `id` | x | | |
| `label` | x | | |
| `description` | x | | |
| `level` | x | | |
| `sortOrder` | x | | |
| `confirm` | x | | |
| `notifications` | x | | |
| `aclProtected` | x | | |
| `path` | | x | |
| `sandboxPermissions` | | x | |
| `runtimeAction` | | | x |
| `timeout` | | | x |

> The `view` and `worker` variants are strict — setting `path` or `sandboxPermissions` on a `worker` button, or `runtimeAction` or `timeout` on a `view` button, fails validation.

Reviewer 2: ```suggestion
The `view` and `worker` variants are strict. Setting `path` or `sandboxPermissions` on a `worker` button, or `runtimeAction` or `timeout` on a `view` button, fails validation.
```

| `label` | string | Yes | On-button text rendered in Admin. |
| `description` | string | No | Summary exposed through `app-config` for installation tooling. |
| `level` | number | No | Position of the button in the toolbar: `-1` (left), `0` (center), or `1` (right). |
| `sortOrder` | number | No | Positive number controlling the button's order within its `level`. |
| `confirm` | object | No | `{ title?, message? }` confirmation dialog shown before the handler runs. |
| `notifications` | object | No | `{ success?, error? }` banner messages shown in Commerce Admin after the handler returns. When omitted, Commerce displays a default success or error banner. |
| `aclProtected` | boolean | No | See [ACL-protected extension points](#acl-protected-extension-points). |
| `path` (`view`) | string | Yes | In-app iframe URL, for example `#/delete-order`. |
| `sandboxPermissions` (`view`) | array | No | One or more of `allow-downloads`, `allow-modals`, `allow-popups`. |
| `runtimeAction` (`worker`) | string | Yes | `package/action` path. Registered as a `workerProcess` operation automatically. |
| `timeout` (`worker`) | number | No | Timeout in seconds. |

### View order view button page

> A `view` button opens an iframe at `<extension-host>/index.html<path>?orderId=<orderId>` inside your App Builder frontend — there's no server-side handler. Read the order ID and close the iframe with the React hooks exported from `@adobe/aio-commerce-lib-admin-ui/web`:

Reviewer 2: ```suggestion
A `view` button opens an iframe at `<extension-host>/index.html<path>?orderId=<orderId>` inside your App Builder frontend. There is no server-side handler. Read the order ID and close the iframe with the React hooks exported from `@adobe/aio-commerce-lib-admin-ui/web`:
```

```jsx
import { useHostConnection, useOrderViewButtonContext } from "@adobe/aio-commerce-lib-admin-ui/web";

export function DeleteOrderPage() {
  const { orderId } = useOrderViewButtonContext();
  const { close } = useHostConnection();

  // ...delete orderId, then call close() to return to the order view.
}
```

### Order view button worker handler

A `worker` button POSTs to your runtime action; parse the request and build the response with `@adobe/aio-commerce-lib-admin-ui/order-view-buttons`:

```js
import {
  okOrderViewButtonResponse,
  parseOrderViewButtonRequest,
} from "@adobe/aio-commerce-lib-admin-ui/order-view-buttons";

export async function main(params) {
  const { id, orderId } = parseOrderViewButtonRequest(params);

  await syncInventory(orderId);
  return okOrderViewButtonResponse();
}
```

Use `orderViewButtonErrorResponse(status, message)` to report a failure. See the [`@adobe/aio-commerce-lib-admin-ui` usage guide](https://github.com/adobe/aio-commerce-sdk/blob/main/packages/aio-commerce-lib-admin-ui/docs/usage.md#order-view-button-wire-contract) for the full wire contract.

## ACL-protected extension points

> Set `aclProtected: true` on a menu, grid column, mass action, or order view button to have Commerce generate a per-app ACL resource for that item and add it to the Adobe Commerce User Roles tree. Admins can then grant or deny the resource per role; users without the resource don't see the item and can't invoke it.

Reviewer 2: ```suggestion
Set `aclProtected: true` on a menu, grid column, mass action, or order view button to have Commerce generate a per-app ACL resource for that item and add it to the Adobe Commerce User Roles tree. Admins can then grant or deny the resource per role. Users without the resource do not see the item and cannot invoke it.
```

For general Admin UI SDK concepts and extension points outside of App Management, see the [Admin UI SDK](../../admin-ui-sdk/index.md) documentation.

## Add a menu declaration

Declare a single Commerce Admin menu entry for the application. Use the named constants exported by `@adobe/aio-commerce-lib-admin-ui/menu` instead of raw strings for `parentMenu`:

```js
import { defineConfig } from "@adobe/aio-commerce-lib-app/config"
import { MENU_CATALOG } from "@adobe/aio-commerce-lib-admin-ui/menu";

export default defineConfig({
  metadata: {
    // ...
  },
  adminUi: {
    menu: {
      id: "approval_dashboard",
      label: "Approval Dashboard",
      description: "Review and approve purchase requests from Commerce Admin.",

>       parentMenu: MENU_CATALOG,

Reviewer: For a purchase approval example (similar to our reference app), maybe the `parentMenu` makes more sense if it is `MENU_SALES`

| `sandboxPermissions` (`view`) | array | No | One or more of `allow-downloads`, `allow-modals`, `allow-popups`. |
| `runtimeAction` (`worker`) | string | Yes | `package/action` path. Registered as a `workerProcess` operation automatically. |
| `timeout` (`worker`) | number | No | Timeout in seconds. |

### View order view button page

A `view` button opens an iframe at `<extension-host>/index.html<path>?orderId=<orderId>` inside your App Builder frontend. There is no server-side handler. Read the order ID and close the iframe with the React hooks exported from `@adobe/aio-commerce-lib-admin-ui/web`:

```jsx
import { useHostConnection, useOrderViewButtonContext } from "@adobe/aio-commerce-lib-admin-ui/web";

export function DeleteOrderPage() {
  const { orderId } = useOrderViewButtonContext();
  const { close } = useHostConnection();

  // ...delete orderId, then call close() to return to the order view.
}
```

### Order view button worker handler

A `worker` button POSTs to your runtime action; parse the request and build the response with `@adobe/aio-commerce-lib-admin-ui/order-view-buttons`:

```js

> import {
>   okOrderViewButtonResponse,
>   parseOrderViewButtonRequest,
> } from "@adobe/aio-commerce-lib-admin-ui/order-view-buttons";
> 
> export async function main(params) {
>   const { id, orderId } = parseOrderViewButtonRequest(params);
> 
>   await syncInventory(orderId);
>   return okOrderViewButtonResponse();
> }

Reviewer: ```suggestion
async function syncInventory(orderId) {
  // Your sync inventory logic for `orderId`
}

export async function main(params) {
  const { id, orderId } = parseOrderViewButtonRequest(params);

  await syncInventory(orderId);
  return okOrderViewButtonResponse();
}
```

  parseMassActionRequest,
} from "@adobe/aio-commerce-lib-admin-ui/mass-actions";

export async function main(params) {
  const { gridType, selectedIds } = parseMassActionRequest(params);

  await approveOrders(selectedIds);
  return okMassActionResponse({ approved: selectedIds.length });
}
```

Use `massActionErrorResponse(status, message)` to report a failure. See the [`@adobe/aio-commerce-lib-admin-ui` usage guide](https://github.com/adobe/aio-commerce-sdk/blob/main/packages/aio-commerce-lib-admin-ui/docs/usage.md#mass-action-worker-contract) for the full wire contract, including the end-to-end ACL example.

### View mass action page

A `view` mass action opens an iframe at `path` inside your App Builder frontend — there's no server-side handler. Read the selected row IDs and close the iframe with the React hooks exported from `@adobe/aio-commerce-lib-admin-ui/web`:

```jsx
import { useHostConnection, useMassActionContext } from "@adobe/aio-commerce-lib-admin-ui/web";

export function ExportOrdersPage() {
  const { selectedIds } = useMassActionContext();
  const { close } = useHostConnection();

  // ...export selectedIds, then call close() to return to the order grid.

> }

Reviewer: This examples need updating to the `data | error` pattern.

| `sandboxPermissions` (`view`) | array | No | One or more of `allow-downloads`, `allow-modals`, `allow-popups`. |
| `runtimeAction` (`worker`) | string | Yes | `package/action` path. Registered as a `workerProcess` operation automatically. |
| `timeout` (`worker`) | number | No | Timeout in seconds. |

### View order view button page

A `view` button opens an iframe at `<extension-host>/index.html<path>?orderId=<orderId>` inside your App Builder frontend. There is no server-side handler. Read the order ID and close the iframe with the React hooks exported from `@adobe/aio-commerce-lib-admin-ui/web`:

```jsx
import { useHostConnection, useOrderViewButtonContext } from "@adobe/aio-commerce-lib-admin-ui/web";

export function DeleteOrderPage() {
  const { orderId } = useOrderViewButtonContext();
  const { close } = useHostConnection();

  // ...delete orderId, then call close() to return to the order view.
}
```

### Order view button worker handler

A `worker` button POSTs to your runtime action; parse the request and build the response with `@adobe/aio-commerce-lib-admin-ui/order-view-buttons`:

```js
import {
  okOrderViewButtonResponse,
  parseOrderViewButtonRequest,

> } from "@adobe/aio-commerce-lib-admin-ui/order-view-buttons";
> 
> async function syncInventory(orderId) {
>   // Your sync inventory logic for `orderId`
> }
> export async function main(params) {
>   const { id, orderId } = parseOrderViewButtonRequest(params);
>   await syncInventory(orderId);
>   return okOrderViewButtonResponse();
> }

Reviewer: Formatting nitpick:
```suggestion
async function syncInventory(orderId) {
  // Your sync inventory logic for orderId
}

export async function main(params) {
  const { id, orderId } = parseOrderViewButtonRequest(params);
  await syncInventory(orderId);
  
  return okOrderViewButtonResponse();
}
```

  metadata: {
    // ...
  },
  adminUi: {
    menu: {
      id: "approval_dashboard",
      label: "Approval Dashboard",
      description: "Review and approve purchase requests from Commerce Admin.",

>       parentMenu: "catalog",

Reviewer: Probably we should use a constant from `lib-admin-ui` in this example

Reviewer 3: can you please provide the example or add a suggestion?

Reviewer: He had already documented it correctly in other parts, so I assumed he knew how it works. @keharper Let me know otherwise.

      sandboxPermissions: ["allow-popups", "allow-downloads"],
      aclProtected: true,
    },
  },
});
```

| Property | Type | Required | Description |
|----------|------|----------|-------------|
| `id` | string | Yes | App-local menu identifier. Allowed characters: letters, digits, `/`, `:`, `_`. |
| `label` | string | Yes | Menu label rendered in Commerce Admin. |
| `description` | string | Yes | Summary shown in installation and permission-review surfaces. |
| `pageTitle` | string | No | Page title for the menu entry. |

> | `parentMenu` | string | No | Existing Commerce menu ID under which the app menu is attached. When omitted, App Management generates a per-app section automatically from `metadata`. |

Reviewer: ```suggestion
| `parentMenu` | string | No | Existing Commerce menu ID under which the app menu is attached. When omitted, App Management generates a per-app section automatically from `metadata` and places it under the `Apps` menu. |
```

Reviewer 3: On this, even when the `parentMenu` is added, we generate a per-app section automatically from `metadata.displayName` and we add it to the specified `parentMenu`

Reviewer: @keharper This should be something in the lines of the below suggestion (feel free to correct wording if it reads odd):

> Existing Commerce menu ID under which the app menu is attached. App Management automatically generates a per-app section from `metadata.displayName`. When omitted, Commerce places it under the `Apps` menu.

        id: "export-orders",
        label: "Export",
        title: "Export orders",
        path: "#/export-orders",
        sandboxPermissions: ["allow-downloads"],
      },
      {
        type: "worker",
        id: "bulk-approve",
        label: "Approve",
        confirm: { message: "Approve the selected orders?" },
        runtimeAction: "orders/bulk-approve",
        timeout: 15,
        notifications: {
          success: "Orders approved.",
          error: "Approval failed. Check the runtime logs.",
        },
      },
    ],
  },
}
```

Mass actions are supported on `order`, `product`, and `customer`. The `id` is authored as a bare name (for example `bulk-approve`); Commerce handles prefixing and collision resolution when rendering the final Admin UI configuration.

### Field applicability by variant

| Field | Common | `view` only | `worker` only |
|-------|:------:|:-----------:|:-------------:|
| `id` | x | | |
| `label` | x | | |
| `title` | x | | |
| `description` | x | | |
| `confirm` | x | | |
| `notifications` | x | | |
| `selectionLimit` | x | | |
| `aclProtected` | x | | |
| `path` | | x | |
| `sandboxPermissions` | | x | |
| `runtimeAction` | | | x |

> | `timeout` | | | x |

Reviewer: I believe this kind of field applicability also exists for view order buttons? Can you confirm @obarcelonap?

Reviewer 4: It is missing for order view buttons in the usage file, I'll add it.
@keharper this pattern of conditional fields is expressed differntly in business config. https://adobedocs.github.io/commerce-extensibility/commerce/extensibility/app-management/configuration-schema#schema-properties
we might want to align the docs in the same pattern

Reviewer: This has been applied, resolving.

The `view` and `worker` variants are strict — setting `path` or `sandboxPermissions` on a `worker` action, or `runtimeAction` or `timeout` on a `view` action, fails validation.

| Property | Type | Required | Description |
|----------|------|----------|-------------|
| `id` | string | Yes | Stable action identifier. |
| `label` | string | Yes | Action label rendered in the Admin UI. |
| `title` | string | No | Page title rendered in the iframe (`view`) or confirmation surface (`worker`). |
| `description` | string | No | Summary exposed through `app-config` for installation tooling. |
| `confirm` | object | No | `{ title?, message? }` confirmation dialog shown before the action runs. |

> | `notifications` | object | No | `{ success?, error? }` toast strings shown after the action completes. |

Reviewer: This is more akin to the old banner notifications, not toast strings.

        label: "Delete",
        description: "Permanently removes the order and its associated records.",
        path: "#/delete-order",
        level: 0,
        sortOrder: 80,
        sandboxPermissions: ["allow-modals", "allow-popups"],
        confirm: { message: "Are you sure you want to delete this order?" },
      },
      {
        type: "worker",
        id: "sync-inventory",
        label: "Sync inventory",
        description: "Pushes the latest stock counts for this order's items to the ERP.",
        runtimeAction: "orders/sync-inventory",
        timeout: 15,
        level: 1,
        sortOrder: 10,
        notifications: {
          success: "Inventory synced successfully.",
          error: "Inventory sync failed. Check the runtime logs.",
        },
      },
    ],
  },
}
```

Order view buttons are only available on `order`.

| Property | Type | Required | Description |
|----------|------|----------|-------------|
| `id` | string | Yes | Stable button identifier, served to Commerce as-is. |
| `label` | string | Yes | On-button text rendered in Admin. |
| `description` | string | No | Summary exposed through `app-config` for installation tooling. |

> | `level` | number | No | `-1`, `0`, or `1`. |

Reviewer: @asalloum5 Can we explain what `level` means?

Reviewer 3: this is a layout placement of buttons within the order view, not sure what exactly it means

| `sortOrder` | number | No | Positive number controlling display order. |
| `confirm` | object | No | `{ title?, message? }` confirmation dialog shown before the handler runs. |

> | `notifications` | object | No | `{ success?, error? }` toast strings shown after the handler returns. |

Reviewer: Idem on `notifications` being banner notifications and not toast strings.

| `runtimeAction` (`worker`) | string | Yes | `package/action` path. Registered as a `workerProcess` operation automatically. |
| `timeout` (`worker`) | number | No | Timeout in seconds. |

A `view` button opens an iframe at `<extension-host>/index.html<path>?orderId=<orderId>` and signals completion back to Commerce through the UIX host connection (`close()` or `onError()`) — no server-side handler is required. A `worker` button POSTs to your runtime action; parse the request and build the response with `@adobe/aio-commerce-lib-admin-ui/order-view-buttons`:

```js
import {
  okOrderViewButtonResponse,
  parseOrderViewButtonRequest,
} from "@adobe/aio-commerce-lib-admin-ui/order-view-buttons";

export async function main(params) {
  const { id, orderId } = parseOrderViewButtonRequest(params);

  await syncInventory(orderId);
  return okOrderViewButtonResponse();
}
```

Use `orderViewButtonErrorResponse(status, message)` to report a failure. See the [`@adobe/aio-commerce-lib-admin-ui` usage guide](https://github.com/adobe/aio-commerce-sdk/blob/main/packages/aio-commerce-lib-admin-ui/docs/usage.md#order-view-button-wire-contract) for the full wire contract.

> ## ACL-protected extension points

Reviewer 3: am not sure about this section, @iivvaannxx do we have these helpers?

Reviewer: Yes, we do, @oshmyheliuk implemented them.

| App root | `getAclResourceId(metadataId)` from `/api` |
| Menu | `getMenuAclResourceId(metadataId, menuId)` from `/menu` |
| Grid column | `getGridColumnAclResourceId(metadataId, entity, columnId)` from `/grid-columns` |
| Mass action | `getMassActionAclResourceId(metadataId, entity, actionId)` from `/mass-actions` |
| Order view button | `getOrderViewButtonAclResourceId(metadataId, buttonId)` from `/order-view-buttons` |

Check the resource from the runtime action handler with `getAdminUiPermissionClient` before serving protected content:

```js
import {
  AdminUiPermissionDeniedError,
  getAdminUiPermissionClient,
} from "@adobe/aio-commerce-lib-admin-ui/api";
import { getMassActionAclResourceId } from "@adobe/aio-commerce-lib-admin-ui/mass-actions";

const permissionClient = getAdminUiPermissionClient({ httpClient, appId: "approval-dashboard-app" });

try {
  await permissionClient.require(
    getMassActionAclResourceId("approval-dashboard-app", "order", "bulk-approve"),
  );
} catch (error) {
  if (error instanceof AdminUiPermissionDeniedError) {
    return massActionErrorResponse(403, "You do not have access to this action");
  }
  throw error;

> }

Reviewer: @oshmyheliuk Could you update the `usage.md` to also show how the `httpClient` is used in these examples created? (I assume the example was taken from our Markdown docs)

Reviewer 5: @iivvaannxx, do you mean an example of how to check the permissions? We have it in usage.md of admin-ui module here https://github.com/adobe/aio-commerce-sdk/blob/main/packages/aio-commerce-lib-admin-ui/docs/usage.md#permission-client.

We can add here to the

Reviewer: No, I meant that this line 

```ts
const permissionClient = getAdminUiPermissionClient({ httpClient, appId: "approval-dashboard-app" });
```

Is using an `httpClient` without showing where it came from or how it was created. Probably it's the context that someone unfamiliar with our libs may find missing. So was referring to augment that example (which I assume was taken from `usage.md`) with that extra code to make it "self-contained" and easily copy-pasteable.

Reviewer 5: Resolved based on discussion with @iivvaannxx

Reviewer: @keharper My bad, this example is already correct in our docs, but in here it's partially incomplete.

```suggestion
import { getCommerceClient } from "@adobe/aio-commerce-lib-app";
import { resolveImsAuthParams } from "@adobe/aio-commerce-lib-auth";

import appConfig from "#app.commerce.config";

export async function main(params) {
  const appId = appConfig.metadata.id;
  const httpClient = await getCommerceClient(resolveImsAuthParams(params));
  const permissionClient = getAdminUiPermissionClient({ httpClient, appId });

  try {
    await permissionClient.require(
      getMassActionAclResourceId(appId, "order", "bulk-approve"),
    );
    
    // Do something after checking permission is granted.
  } catch (error) {
    if (error instanceof AdminUiPermissionDeniedError) {
      return massActionErrorResponse(403, "You do not have access to this action");
    }
    throw error;
  }
}
```
